src/main/java/uk/gov/justice/digital/hmpps/whereabouts/controllers/ControllerAdvice.kt

```diff
@@ -0,0 +1,161 @@
+package uk.gov.justice.digital.hmpps.whereabouts.controllers
+
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.http.MediaType
+import org.springframework.http.ResponseEntity
+import org.springframework.security.access.AccessDeniedException
+import org.springframework.web.bind.annotation.ExceptionHandler
+import org.springframework.web.bind.annotation.RestControllerAdvice
+import org.springframework.web.client.RestClientException
+import org.springframework.web.client.RestClientResponseException
+import org.springframework.web.reactive.function.client.WebClientResponseException
+import uk.gov.justice.digital.hmpps.whereabouts.dto.ErrorResponse
+import uk.gov.justice.digital.hmpps.whereabouts.services.AttendanceExists
+import uk.gov.justice.digital.hmpps.whereabouts.services.AttendanceLocked
+import uk.gov.justice.digital.hmpps.whereabouts.services.InvalidCourtLocation
+import uk.gov.justice.digital.hmpps.whereabouts.services.ValidationException
+import javax.persistence.EntityNotFoundException
+
+@RestControllerAdvice(basePackageClasses = [AttendanceController::class, AttendanceStatisticsController::class, AttendancesController::class, CourtController::class, AgencyController::class, LocationController::class, CellMoveController::class])
+class ControllerAdvice {
+  @ExceptionHandler(RestClientResponseException::class)
+  fun handleException(e: RestClientResponseException): ResponseEntity<ByteArray> {
+    log.error("Unexpected exception", e)
+    return ResponseEntity
+      .status(e.rawStatusCode)
+      .body(e.responseBodyAsByteArray)
+  }
+
+  @ExceptionHandler(RestClientException::class)
+  fun handleException(e: RestClientException): ResponseEntity<ErrorResponse> {
+    log.error("Unexpected exception", e)
+    return ResponseEntity
+      .status(HttpStatus.INTERNAL_SERVER_ERROR)
+      .body(
+        ErrorResponse
+          .builder()
+          .status(HttpStatus.INTERNAL_SERVER_ERROR.value())
+          .developerMessage(e.message)
+          .build()
+      )
+  }
+
+  @ExceptionHandler(WebClientResponseException::class)
+  fun handleException(e: WebClientResponseException): ResponseEntity<ErrorResponse> {
+    log.error("Unexpected exception", e)
+    return ResponseEntity
+      .status(HttpStatus.INTERNAL_SERVER_ERROR)
+      .body(
+        ErrorResponse
+          .builder()
+          .status(HttpStatus.INTERNAL_SERVER_ERROR.value())
+          .developerMessage(e.message)
+          .build()
+      )
+  }
+
+  @ExceptionHandler(AccessDeniedException::class)
+  fun handleException(e: AccessDeniedException?): ResponseEntity<ErrorResponse> {
+    log.debug("Forbidden (403) returned", e)
```

Comment:
shouldn't really log `e` here, maybe just the message, otherwise appears in app insights in the `exceptions` tab.  Same with the rest of file, only log `e` if it is a server error